tests/worker/test_workflow.py

```diff
@@ -789,6 +899,255 @@ async def test_workflow_cancel_multi(client: Client):
+            workflow.execute_activity(
+                wait_cancel, schedule_to_close_timeout=timedelta(seconds=5)
+            )
+        )
+        activity_task.cancel()
+        await self.wait_and_swallow(activity_task)
+
+        # Start local activity
+        activity_handle = workflow.start_local_activity(
+            wait_cancel, schedule_to_close_timeout=timedelta(seconds=5)
+        )
+        activity_handle.cancel()
+        await self.wait_and_swallow(activity_handle)
+
+        # Execute local activity
+        activity_task = asyncio.create_task(
+            workflow.execute_local_activity(
+                wait_cancel, schedule_to_close_timeout=timedelta(seconds=5)
+            )
+        )
+        activity_task.cancel()
+        await self.wait_and_swallow(activity_task)
+
+        # Start child
+        child_task1 = asyncio.create_task(
+            workflow.start_child_workflow(
+                LongSleepWorkflow.run, id=f"{workflow.info().workflow_id}_child1"
+            )
+        )
+        child_task1.cancel()
+        await self.wait_and_swallow(child_task1)
+
+        # Execute child
+        child_task2 = asyncio.create_task(
+            workflow.execute_child_workflow(
+                LongSleepWorkflow.run, id=f"{workflow.info().workflow_id}_child2"
+            )
+        )
+        child_task2.cancel()
+        await self.wait_and_swallow(child_task2)
+
+        # Sleep for a short bit to force another task to run so we know that
+        # workflow completion isn't saving us here
+        await asyncio.sleep(0.01)
+
+    async def wait_and_swallow(self, aw: Awaitable) -> None:
+        try:
+            await aw
+        except:
+            pass
+
+
+async def test_workflow_cancel_unsent(client: Client):
+    workflow_id = f"workflow-{uuid.uuid4()}"
+    async with new_worker(
+        client, CancelUnsentWorkflow, LongSleepWorkflow, activities=[wait_cancel]
+    ) as worker:
+        await client.execute_workflow(
+            CancelUnsentWorkflow.run,
+            id=workflow_id,
+            task_queue=worker.task_queue,
+        )
+    # Check history
+    resp = await client.service.get_workflow_execution_history(
+        GetWorkflowExecutionHistoryRequest(
+            namespace=client.namespace,
+            execution=WorkflowExecution(workflow_id=workflow_id),
+        )
+    )
+    found_timer = False
+    for event in resp.history.events:
+        # No activities or children scheduled
+        assert event.event_type is not EventType.EVENT_TYPE_ACTIVITY_TASK_SCHEDULED
+        assert (
+            event.event_type
+            is not EventType.EVENT_TYPE_START_CHILD_WORKFLOW_EXECUTION_INITIATED
+        )
+        # Make sure timer is just our 0.01 timer
+        if event.event_type is EventType.EVENT_TYPE_TIMER_STARTED:
+            assert (
+                event.timer_started_event_attributes.start_to_fire_timeout.ToMilliseconds()
+                == 10
```

Comment:
Hmm.. looking at this now I'm wondering if it's worth not sending this command if it's cancelled in the same activation. (Would be a backwards incompatible Core change).

Reply:
From what I saw in history, it seems core properly handles cancelling out a command sent in the same activation completion. It's for that reason I didn't add extra code to remove commands of things I cancel in the same completion before sending the completion - it seems like core does it.